### DISCUSSION TITLE
Sparta - Checkpointing data isn't derived from sparta::ArchData

### DISCUSSION THREAD

**Discussion by abdu-neubla** [sparcians/map#424] [Q&A] [answered]
Hi,

I'm trying to checkpoint some data/class that isn't derived from sparta::ArchData. I would like to do it via sparta if it is supported.

I did take a look at this page for info https://sparcians.github.io/map/checkpoint_format.html 

It wasn't clear to me how it is done for non-sparta::ArchData from reading that.

Apologies if I wasn't clear, I'm new to using sparta so I have a lot to learn :)

Thanks,

**Comment by ghost** [comment, marked as answer]
No need to apologies -- Sparta has a bit to go with documentation and clarity on functionality.  Thanks for posting your question.

You are correct that out-of-the-box Sparta will provide basic checkpointing for `sparta::ArchData` objects, but it's not clear/clean on how to checkpoint an object that _isn't_ using an `ArchData` object as the backend storage model.  In fact, it's can be hard to do so.

Here's what we've done in the past, but there are definitely other ways of doing it.  Keep in mind that these methods assume that you are keeping checkpoints in memory and possibly saving them to disk most likely at the end of simulation.

### Writing your own Checkpointer

In this method, the framework will handle any `sparta::ArchData` objects, but it's up to the modeler to checkpoint his/her own structures that are not `sparta::ArchData` objects.

In this method, start with a custom class that responds to a simple API like the one below and contains a `sparta::serialization::checkpoint::PersistentFastCheckpointer` class documented [here](https://sparcians.github.io/map/classsparta_1_1serialization_1_1checkpoint_1_1PersistentFastCheckpointer.html).
```
class MyCheckpointer
{
public:
    MyCheckpointer(sparta::TreeNode * root) :
        arch_based_checkpoints_(*root)
    {}
    uint64_t createCheckpoint() { 
         auto chpt_id = arch_state_checkpoints_.createCheckpoint();
         for(auto acp : additional_non_archdata_checkpointables_) {
             acp->createCheckpoint(chpt_id);
         }
         return chpt_id;
    }
    void restoreCheckpoint(uint64_t chpt_id) { ... }
    void deleteCheckpoint(uint64_t chpt_id) {...}
    void createHead() { ... }
private:
    using PersistentFastCheckpointer = sparta::serialization::checkpoint::PersistentFastCheckpointer;
    PersistentFastCheckpointer arch_based_checkpoints_;

    std::list<Checkpointable*> additional_non_archdata_checkpointables_;
};
```
In your implementation of the class above, you can "walk" registered `Checklpointable` objects and save state to a block of allocated memory not controlled by the framework. 

For example, a `Checkpointable` might have an API that looks like this:

```
class Checkpointable
{
public:
        template<class ObjT>
        void registerForCheckpointing(ObjT * obj)
        {
            static_assert(std::is_trivially_copyable_v<ObjT> || std::is_assignable_v<ObjT, ObjT>,
                          "Object is not copyable nor assignable.  Cannot checkpoint it");
            objs_to_chlpt_.emplace_back(new ObjectBlockStorage<ObjT>(obj));
        }
 
      void createCheckpoint(...) { ... }
      void restoreCheckpoint(...) { ... }
      // ... etc
};
```

And if you have a class that needs to be checkpointed, you'd do this:
```
class MySimClass : public Checkpointable
{
public:
    MySimClass(...) {
         registerForCheckpointing(&my_state_to_be_saved_);
    }

private: 
    uint32_t my_state_to_be_saved_ = 0;
};
```

It's up to you how you store the objects in memory and eventually to disk once the [save method](https://sparcians.github.io/map/classsparta_1_1serialization_1_1checkpoint_1_1PersistentFastCheckpointer.html#ad1bd4c5d578dc25430c5106e7f04391d) is called.

### Saving all state in sparta::ArchData

Another way to save state is to have your local variables use the `sparta::ArchData` as it's memory.  This method is some C++ trickery, but can be done.  I'll admit it's been a while since I tried it, so I probably have a few things wrong.

To get started, created a `sparta::ArchData` with a given line size that should fit the data you wish to store/checkpoint.  
```
   sparta::ArchData arch_data_{my_node, 1024};
```
Get the `sparta::ArchData::Line` contained in the block and set it dirty (so it will be check-pointed):
```
    auto line = arch_data_.getLine(0);
    line->flagDirty();
```
Now, map local variables into that memory using an in-place `new`:
```
    uint32_t * my_local_var_ = new (line) uint32_t(0);
    line += sizeof(uint32_t);
    uint32_t * my_other_local_var_ = new (line) uint32_t(0);
```

Yes, evil, I know.  But now you can use the `ArchData` to save/restore your local variables and not have to write a boat-load of checkpointing code.  ;)

**Reply by abdu-neubla** [reply]
Thanks Knute 😄 I will probably end up going with writing my own checkpointer. seems easier to work with if lets say that class I'm checkpointing has changed. And I also worry that saving all state in sparta::ArchData would be harder to be debug.
